core/services/functions/connector_handler.go

```diff
@@ -160,6 +200,105 @@ func (h *functionsConnectorHandler) handleSecretsSet(ctx context.Context, gatewa
 	h.sendResponseAndLog(ctx, gatewayId, body, response)
 }
 
+func (h *functionsConnectorHandler) handleHeartbeat(ctx context.Context, gatewayId string, requestBody *api.MessageBody, fromAddr ethCommon.Address) {
+	var request *OffchainRequest
+	err := json.Unmarshal(requestBody.Payload, &request)
+	if err != nil {
+		h.sendResponseAndLog(ctx, gatewayId, requestBody, HeartbeatResponse{
+			Status:        RequestStateInternalError,
+			InternalError: fmt.Sprintf("failed to unmarshal request: %v", err),
+		})
+		return
+	}
+	if !bytes.Equal(request.RequestInitiator, fromAddr.Bytes()) {
+		h.sendResponseAndLog(ctx, gatewayId, requestBody, HeartbeatResponse{
+			Status:        RequestStateInternalError,
+			InternalError: "requestInitiator doesn't match sender",
+		})
+		return
+	}
+
+	internalId := InternalId(fromAddr.Bytes(), request.RequestId)
```

Comment:
If I get this right, the only purpose of this "internal ID" is to make it the key to `map`.
However, you could simply make the key to be a value struct,
```
type heartbeatRequestKey struct {
  sender []byte
  requestId RequestID
}
```
and then this struct can be directly used by map which will does hashing internally.
This would look much clearer to me.

Reply:
We still need to have a single ID for request processing in the core node.

Reply:
Can't it be just RequestID?

Reply:
No because in the offchain scenario, RequestID is chosen by a user so we can't guarantee global uniqueness.